src/dotnet/GeneratorAgent/src/Azure.Tools.GeneratorAgent/GithubFilesRetrieveService.cs

```diff
@@ -0,0 +1,152 @@
+
+                HttpResponseMessage response = await HttpClient.GetAsync(apiUrl, cancellationToken).ConfigureAwait(false);
+                if (!response.IsSuccessStatusCode)
+                {
+                    Logger.LogWarning("GitHub API request failed: {StatusCode} {ReasonPhrase}", 
+                        response.StatusCode, response.ReasonPhrase);
+                    throw new HttpRequestException($"GitHub API request failed: {response.StatusCode} {response.ReasonPhrase}");
+                }
+
+                string jsonContent = await response.Content.ReadAsStringAsync(cancellationToken).ConfigureAwait(false);
+                GitHubContent[]? contents = JsonSerializer.Deserialize<GitHubContent[]>(jsonContent, JsonOptions);
+
+                if (contents == null)
+                {
+                    throw new InvalidOperationException("Failed to deserialize GitHub API response");
+                }
+
+                Dictionary<string, string> typeSpecFiles = new(contents.Length);
+
+                IEnumerable<GitHubContent> tspFiles = contents.Where(c => 
+                    string.Equals(c.Type, "file", StringComparison.Ordinal) && 
+                    c.Name.EndsWith(".tsp", StringComparison.OrdinalIgnoreCase));
+
+                // Download each .tsp file content in parallel
+                Task<(string FileName, string Content)>[] downloadTasks = tspFiles
+                    .Where(tspFile => !string.IsNullOrEmpty(tspFile.DownloadUrl))
+                    .Select(tspFile => DownloadFileContentAsync(tspFile.Name, tspFile.DownloadUrl!, cancellationToken))
+                    .ToArray();
+
+                if (downloadTasks.Length == 0)
+                {
+                    Logger.LogWarning("No valid .tsp files found with download URLs");
+                    throw new InvalidOperationException("No valid .tsp files found with download URLs");
+                }
+
+                (string FileName, string Content)[] downloadedFiles = await Task.WhenAll(downloadTasks).ConfigureAwait(false);
+
+                foreach ((string fileName, string content) in downloadedFiles)
+                {
+                    if (!string.IsNullOrEmpty(content))
+                    {
+                        typeSpecFiles[fileName] = content;
+                        Logger.LogDebug("Downloaded file: {FileName} ({Size} characters)", fileName, content.Length);
+                    }
+                }
+
+                return Result<Dictionary<string, string>>.Success(typeSpecFiles);
+            }
+            catch (Exception ex)
+            {
+                Logger.LogCritical(ex, "Error in GitHub API TypeSpec files fetch");
+                throw;
+            }
+        }
+
+        private async Task<(string FileName, string Content)> DownloadFileContentAsync(string fileName, string downloadUrl, CancellationToken cancellationToken)
```

Comment:
nit but if it's ok for downloadUrl to be null it's nice to be more explicit and have this be string? and then you can remove the null forgiving operator in L96